### PR TITLE
spirv-fuzz: Use component-wise selectors when flattening conditional branches

### LINKED ISSUE #3920: spirv-fuzz: Problem with phi-to-select for older SPIR-V versions
Pre SPIR-V 1.4 it was not legal to perform OpSelect on non scalar/vector/pointer types, and the vector case was restricted to working in a component-wise fashion.

TransformationFlattenConditionalBranch assumes that `OpSelect %type %boolean_id %v1 %v2` will always work, but pre SPIR-V 1.4 this may not be allowed at all, or might require `%boolean_id` to actually be a boolean vector.

### PR DESCRIPTION
**Pull request by afd**
Fixes #3920.

### CODE REVIEW

File: source/fuzz/transformation_flatten_conditional_branch.cpp [hunk 1]
```diff
@@ -52,6 +57,63 @@ bool TransformationFlattenConditionalBranch::IsApplicable(
...
+        message_.fresh_id_for_bvec4_selector()}) {
+    if (fresh_id_for_bvec_selector != 0) {
+      if (!CheckIdIsFreshAndNotUsedByThisTransformation(
+              fresh_id_for_bvec_selector, ir_context, &used_fresh_ids)) {
+        return false;
+      }
+    }
+  }
+
+  if (OpSelectArgumentsAreRestricted(ir_context)) {
```
Comment (paulthomson) [inline source/fuzz/transformation_flatten_conditional_branch.cpp:76]: Could add a comment to describe what this next statement is doing.

File: source/fuzz/transformation_flatten_conditional_branch.cpp [hunk 2]
```diff
@@ -734,5 +744,165 @@ TransformationFlattenConditionalBranch::GetFreshIds() const {
...
+       &require_4d_boolean_vector](opt::Instruction* phi_inst) {
+        assert(phi_inst->NumInOperands() == 4 &&
+               "We are going to replace an OpPhi with an OpSelect.  This "
+               "only makes sense if the block has two distinct "
+               "predecessors.");
+        // The OpPhi takes values from two distinct predecessors.  One
+        // predecessor is associated with the "true" path of the conditional
+        // we are flattening, the other with the "false" path, but these
+        // predecessors can appear in either order as operands to the OpPhi
+        // instruction.
```
Comment (paulthomson) [inline source/fuzz/transformation_flatten_conditional_branch.cpp:822]: This comment could be moved down to where we use the dominator analysis. We could also mention that we achieve this by checking dominance of the true and false immediate successors of the header block.

**Review by paulthomson — APPROVED** [review]

File: source/fuzz/fuzzer_pass_flatten_conditional_branches.cpp [hunk 3]
```diff
@@ -72,6 +72,106 @@ void FuzzerPassFlattenConditionalBranches::Apply() {
...
+                        &fresh_id_for_bvec2_selector);
+                    break;
+                  case 3:
+                    PrepareForOpPhiOnVectors(
+                        dimension,
+                        use_component_wise_3d_select_even_if_optional,
+                        &fresh_id_for_bvec3_selector);
+                    break;
+                  default:
+                    assert(dimension == 4 && "Invalid vector dimension.");
```
[the remark this answers is not in the document]
Comment (afd) [reply source/fuzz/fuzzer_pass_flatten_conditional_branches.cpp:161]: Done.

File: source/fuzz/transformation_flatten_conditional_branch.cpp (hunk 1 above)
Comment (afd) [reply source/fuzz/transformation_flatten_conditional_branch.cpp:76]: Done - and I commented that code a bit more in general.

File: source/fuzz/transformation_flatten_conditional_branch.cpp (hunk 2 above)
Comment (afd) [reply source/fuzz/transformation_flatten_conditional_branch.cpp:822]: Done.

**Pull request merged**